### ISSUE TITLE
Interruption of water by waving water shaderㅤ

### ISSUE THREAD

**Issue by Rui-Minetest** [minetest/minetest#3075]

**Comment by est31** [comment]
Possible dupe of  #1404.

**Comment by PilzAdam** [comment]
@est31 no it's not.

Usually the top tile of the water is not visible when a solid node is above it, so it's not added to the mesh. This needs to be changed when the waving shader is enabled (there was a similar thing with new_style_water a while ago).

**Comment by est31** [comment]
I see, that makes sense.

**Comment by RobertZenz** [comment]
@PilzAdam Uh, I've been seeing the exact opposite, the top tile of water is visible when a solid node is above it (which is especially visible in underwater caves for example). Doesn't matter if new_style or not.

**Comment by Rui914** [comment]
https://github.com/minetest/minetest/blob/master/src/nodedef.cpp#L845
Can it is resolve by change this to 0?

**Comment by PilzAdam** [comment]
@Rui914 yes, probably; the if should contain `or wavinShaderEnabled` or something like that.

**Comment by Rui914** [comment]
```
if (g_settings->getBool("enable_waving_water"))
	f->solidness = 0;
else
	f->solidness = 1;
f->backface_culling = false;
```
Is this useless?

**Comment by paramat** [comment]
Waving water will probably be removed soon due to its problems.

**Comment by C1ffisme** [comment]
@paramat Erm... Or we could just try and fix it, maybe? I think it's worth a look into the rendering code...

I imagine that to save extra rendering, Minetest doesn't render the top water face, because normally it would be squashed against the stone on top. But this is what I think, and I might be very wrong.

**Comment by est31** [comment]
Now even I think that it should be removed. its far too buggy.

**Comment by paramat** [comment]
The title should not have been removed, this should not have been closed.

**Comment by Wuzzy2** [comment]
To anyone who has no clue what this bug is about:

## Summary

If a solid block is directly above a waving liquid, and waving liquids are enabled, there's a render bug. The area below the solid block has no visible liquid, thus creating an odd “hole” in the liquid. See screenshot in 1st post.

## Steps to reproduce

0. Enable waving liquids
1. Start MTG
2. Take any solid block like dirt
3. Place the block so that it is directly above water or lava
4. Look at the dirt and the area below

**Comment by Wuzzy2** [comment]
This bug is at least "medium priority" imho (not "low priority") because it is quite visible to players and not rare at all. All you need to do is enable waving liquids and then you can see it happen wherever water is below a solid block. This is common in caves and at cliffs, which you an also see in the screenshot.

**Comment by nininik0** [comment]
I can confirm that the same happens to me especially when there is a floating house on water

**Comment by sfence** [comment]
![screenshot_20240127_071142](https://github.com/minetest/minetest/assets/17455197/54240abf-9517-436e-94ab-52cc2b2a70db)
Is it acceptable to have waving water on the roof of sunken caves also?

**Comment by appgurueu** [comment]
> I've got an idea, probably I'll start a pr these days if it works during the test.

It should be noted that this is definitely not a beginner-friendly bug to fix; there's a reason this bug has been around for a while.

At least two PRs have failed, another one is in a state where it would introduce other bugs.

I would recommend looking at the existing PRs (and why they failed) first before attempting to fix this.

**Comment by appgurueu** [comment]
> > > I've got an idea, probably I'll start a pr these days if it works during the test.
> > 
> > 
> > It should be noted that this is definitely not a beginner-friendly bug to fix; there's a reason this bug has been around for a while.
> > At least two PRs have failed, another one is in a state where it would introduce other bugs.
> > I would recommend looking at the existing PRs (and why they failed) first before attempting to fix this.
> 
> No, I don't think I'm pretty much a beginner

Great, then whether this issue is beginner-friendly or not wouldn't matter to you. 

> nor do I think considering a participant is not good enough for a issue thus talk him out of it is kind of a good action

I don't know whether you are "good enough" for this issue and it is not up to me to decide.

Whether you think you want to tackle this is up to you, and I was merely supplying you with what I consider important information for you to take into account since you announced that you would want to work on this.

> Surely there's people with better techniques on this, but if that means once a PR I start could fail and not merged then I shouldn't try it in the first place, sorry but I disagree with that.

You can work on whatever you like, but if you want to submit a PR to get it fixed for everyone, reviewer time needs to be taken into account too.

It is better to tell potential contributors how to approach an issue properly (including estimating the work required), rather than to reject (possibly after lengthy discussion) deeply flawed PRs resulting from an improper approach. This saves both the contributors and reviewers time and is what I was doing here.

> Mt is an open source community, and in my opinion writing code is about having fun and the participants like me mostly have volunteer intentions. It's not the outcome of participation that really matters, it's the act of trying to contribute to the community to the best of one's ability, isn't it?

Both matter. "Together we serve the users." What's "objectively" best for the project is ultimately paramount, which means spending contributor and reviewer time wisely.

If you want to maximize the fun you have while coding, Minetest is not ideal. I usually have a few "fun" side projects for this purpose.

**Comment by Desour** [comment]
To properly fix this, can we maybe use a new vertex format, where we add a factor of how much the vertex may wave?

So, if the neighboring node has no top face, we'd set the factor for the vertices towards this neighbor to 0.

(The result might look like the water is sticking to the solid node. To fix this, one could make it liquids wave even if there's a solid node above, if there's waving liquid next to the node.)

This vertex format could also be used to fix flowing liquid waving into the ground, and waving leaves waving into solid nodes.

**Comment by nininik0** [comment]
Wasn't this already fixed?

**Comment by sfence** [comment]
No.

**Comment by Wuzzy2** [comment]
#17302 appears to fix this, but the 5.17.0 proposed <https://github.com/SmallJoker/fork-dev.luanti.org/blob/pr_344_changelog_5.17.0/content/about/changelog.md> changelog says:

> Mostly fix the appearance of waving liquids (shader) below solid nodes (ZenonSeth)

I’m not sure why only "mostly".

**Comment by SmallJoker** [comment]
> I’m not sure why only "mostly".

https://github.com/luanti-org/luanti/pull/17302#discussion_r3608225376
